### PR TITLE
Add tomcat dbcp2 custom connection-properties support

### LINKED ISSUE #3384: Implement DataSourcePropertiesSetter by SPI
Some of connection pools can set custom datasource properties in spring, like HikariCP, see #3365 .
Now, setting datasource properties of HikariCP for sharding-jdbc-spring has already implemented.
But there are many of connection pools which can set datasource properties by their own ways.

So I think ShardingSphere should implement DataSourcePropertiesSetter by SPI, make the feature can be expanded easier.

- [x] Abstract DataSourcePropertiesSetter to SPI
- [x] Support frequent connetion poll in sharding-jdbc-spring
     - [x] DBCP2
- [x] Support frequent connetion poll in sharding-jdbc-orchestration-spring
     - [x] HikariCP
     - [x] DBCP2

### PR DESCRIPTION
**Pull request by yanickxia**
Fixes #3384.

Changes proposed in this pull request:
- Add tomcat dbcp2 custom connection-properties support

---
This change is [Reviewable](https://reviewable.io/reviews/apache/incubator-shardingsphere/3392)

### CODE REVIEW

File: sharding-spring/sharding-spring-boot-util/src/main/java/org/apache/shardingsphere/spring/boot/datasource/AbstractDbcp2DataSourcePropertiesSetter.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     * @param dataSourceName current database name
+     * @param dataSource dataSource instance
+     */
+    @SneakyThrows(ReflectiveOperationException.class)
+    public void propertiesSet(final Environment environment, final String prefix, final String dataSourceName, final DataSource dataSource) {
+        String datasourcePropertiesPrefix = prefix + dataSourceName.trim() + ".connection-properties";
+        if (PropertyUtil.containPropertyPrefix(environment, datasourcePropertiesPrefix)) {
+            Map datasourceProperties = PropertyUtil.handle(environment, datasourcePropertiesPrefix, Map.class);
+            Method method = dataSource.getClass().getMethod("addConnectionProperty", String.class, String.class);
+            for (Object key : datasourceProperties.keySet()) {
```
Comment (KomachiSion) [inline sharding-spring/sharding-spring-boot-util/src/main/java/org/apache/shardingsphere/spring/boot/datasource/AbstractDbcp2DataSourcePropertiesSetter.java]: use `each` as the item in for-each loop.

File: sharding-spring/sharding-spring-boot-util/src/test/java/org/apache/shardingsphere/spring/boot/datasource/Dbcp2DataSourcePropertiesSetterTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.test.util.ReflectionTestUtils;
+
+import java.util.Properties;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public class Dbcp2DataSourcePropertiesSetterTest {
+    
+    private final CommonDbcp2DataSourcePropertiesSetter abstractDbcp2DataSourcePropertiesSetter = new CommonDbcp2DataSourcePropertiesSetter();
```
Comment (KomachiSion) [inline sharding-spring/sharding-spring-boot-util/src/test/java/org/apache/shardingsphere/spring/boot/datasource/Dbcp2DataSourcePropertiesSetterTest.java]: Is `dbcp2DataSourcePropertiesSetter` better?

**Comment by KomachiSion** [comment]
CI restart has passed.
https://builds.apache.org/blue/organizations/jenkins/shardingsphere-ci/detail/shardingsphere-ci/764/pipeline

**Comment by terrymanu** [comment]
I just close and reopen this issue for ci failure. It seems like test coverage website broken, let's try ci again.

**Pull request merged**